cirq-core/cirq/sim/simulator.py

```diff
@@ -818,7 +818,15 @@ def __init__(
     ) -> None:
         self.params = params
         self.measurements = measurements
-        self._final_simulator_state = final_simulator_state
+        self._step_result_or_state = final_simulator_state
```

Comment:
The constructor API is fine, but I think it'd be cleaner to identify the type of `final_simulator_state` and store `_step_result` and `_simulator_state` separately.